### PR TITLE
refactor LDAP driver in userprovider

### PR DESCRIPTION
**Pull request by rhafer**
Move some common code into helper functions. Also don't assume that the
userid in reva matches the user's username. So add an additional LDAP
lookup to get the username before looking up groupmembership.

### CODE REVIEW

File: internal/grpc/services/userprovider/userprovider.go [hunk 1]
```diff
@@ -127,10 +127,12 @@ func (s *service) Register(ss *grpc.Server) {
 func (s *service) GetUser(ctx context.Context, req *userpb.GetUserRequest) (*userpb.GetUserResponse, error) {
 	user, err := s.usermgr.GetUser(ctx, req.UserId)
 	if err != nil {
-		// TODO(labkode): check for not found.
-		err = errors.Wrap(err, "userprovidersvc: error getting user")
-		res := &userpb.GetUserResponse{
-			Status: status.NewInternal(ctx, err, "error getting user"),
+		res := &userpb.GetUserResponse{}
+		if _, ok := err.(errtypes.NotFound); ok {
```
[the remark this answers is not in the document]
Comment (rhafer) [reply internal/grpc/services/userprovider/userprovider.go:131]: The specific errortype we're checking for here (errtypes.NotFound) is not wrapped. It's explicitly returned by the respective user driver (see e.g. https://github.com/cs3org/reva/pull/2133/files#diff-c8809eb091bec2149ed081690ddfb7567dfd3940293319f037d0df7b41d5fedfR371 or https://github.com/cs3org/reva/blob/master/pkg/user/manager/demo/demo.go#L63)

File: internal/http/services/owncloud/ocdav/webdav.go [hunk 2]
```diff
@@ -91,7 +94,16 @@ func (h *WebDavHandler) init(ns string, useLoggedInUserNS bool) error {
 // Handler handles requests
 func (h *WebDavHandler) Handler(s *svc) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ns := applyLayout(r.Context(), h.namespace, h.useLoggedInUserNS, r.URL.Path)
+		ns, newPath, err := s.ApplyLayout(r.Context(), h.namespace, h.useLoggedInUserNS, r.URL.Path)
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
```
[the remark this answers is not in the document]
Comment (rhafer) [reply internal/http/services/owncloud/ocdav/webdav.go:99]: ApplyLayout() might certainly return other errors (e.g. when it fails to get a GatewayClient). I am unsure though if we really need to return something else that `StatusNotFound` to the client in that case. @butonic  @aduffeck  any thoughts?

File: pkg/user/manager/ldap/ldap.go [hunk 3]
```diff
@@ -361,16 +337,54 @@ func (m *manager) GetUserGroups(ctx context.Context, uid *userpb.UserId) ([]stri
...
+		filter,
+		[]string{m.c.Schema.DN, m.c.Schema.UID, m.c.Schema.CN, m.c.Schema.Mail, m.c.Schema.DisplayName, m.c.Schema.UIDNumber, m.c.Schema.GIDNumber},
+		nil,
+	)
+
+	sr, err := conn.Search(searchRequest)
+	if err != nil {
+		if lerr, ok := err.(*ldap.Error); ok {
+			if lerr.ResultCode == ldap.LDAPResultSizeLimitExceeded {
+				log.Error().Err(lerr).Msg(fmt.Sprintf("too many results for user filter '%s'", filter))
```
[the remark this answers is not in the document]
Comment (rhafer) [reply pkg/user/manager/ldap/ldap.go:364]: IMO `NotFound` is still ok here. Especially when looking at the other types defined in `errtypes`. I guess we could return a more helpful error message with it though. I'll add that.

File: internal/grpc/services/userprovider/userprovider.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhafer) [reply internal/grpc/services/userprovider/userprovider.go:131]: :+1:

File: internal/http/services/owncloud/ocdav/ocdav.go [hunk 4]
```diff
@@ -235,20 +236,59 @@ func (s *svc) getClient() (gateway.GatewayAPIClient, error) {
...
+		var requestUsernameOrID string
+		requestUsernameOrID, requestPath = router.ShiftPath(requestPath)
+
+		gatewayClient, err := s.getClient()
+		if err != nil {
+			return "", "", err
+		}
+
+		// Check if we got a Username
+		res, err := gatewayClient.GetUserByClaim(ctx, &userpb.GetUserByClaimRequest{
```
Comment (ishank011) [inline internal/http/services/owncloud/ocdav/ocdav.go:260]: These calls can be avoided by checking what is being specified in the template. If we know that the calls will always be to the `/dav/files/$username/$path` and the namespace template asks for username only, then we can return early. Also, these need to be cached. Going to the userprovider service for every storage operation would be costly.

File: internal/http/services/owncloud/ocdav/webdav.go [hunk 5]
```diff
@@ -91,7 +94,16 @@ func (h *WebDavHandler) init(ns string, useLoggedInUserNS bool) error {
...
+			b, err := Marshal(exception{
+				code:    SabredavNotFound,
+				message: fmt.Sprintf("could not get storage for %s", r.URL.Path),
+			})
+			HandleWebdavError(appctx.GetLogger(r.Context()), w, b, err)
+		}
+		r.URL.Path = newPath
```
Comment (ishank011) [inline internal/http/services/owncloud/ocdav/webdav.go:106]: Nice! But this shouldn't have worked previously. I'd recommend having integration tests for both scenarios `files_namespace = "/users/{{.Id.OpaqueId}}"` and `files_namespace = "/users"`

File: pkg/user/manager/ldap/ldap.go [hunk 6]
```diff
@@ -405,7 +420,7 @@ func (m *manager) getFindFilter(query string) string {
 	return strings.ReplaceAll(m.c.FindFilter, "{{query}}", ldap.EscapeFilter(query))
 }
 
-func (m *manager) getGroupFilter(uid *userpb.UserId) string {
+func (m *manager) getGroupFilter(uid interface{}) string {
```
Comment (ishank011) [inline pkg/user/manager/ldap/ldap.go:423]: Can we go back to `uid` being `*userpb.UserId`?

**Comment by rhafer** [comment]
This has landed in edge a while ago. I am going to close this for now. (Trying to bring back the edge an master branch in sync with another PR)

**Pull request closed without merging**